### ISSUE TITLE
Python APIs to call  Losses/Metrics functions implemented in LGB for general calculation purpose

### ISSUE THREAD

**Issue by wenmin-wu** [microsoft/LightGBM#3744]
## Summary
As I mentioned in this [Kaggle discussion](https://www.kaggle.com/c/riiid-test-answer-prediction/discussion/208031), fast implementations of metrics/losses calculation are very essential in many cases. However, the implementation of other packages is not as fast as LGB. It would be great if LGB can provide Python/R APIs for the calculation of these Metrics/Losses.

## Motivation
Leverage the fast implementations of Metrics/Losses in LGB and provide Python/R APIs for general calculation purposes, just like the `sklearn.metrics.roc_auc_score`. It would save us quite a lot of time.

**Comment by StrikerRUS** [comment]
Just to clarify: is your feature request about providing public API for metrics/loses to use in third-party packages?

**Comment by wenmin-wu** [comment]
> Just to clarify: is your feature request about providing public API for metrics/loses to use in third-party packages?

Yup @StrikerRUS , maybe LGB can decouple the metrics/losses to small lib

**Comment by StrikerRUS** [comment]
Thanks for elaborating, got it!

**Comment by StrikerRUS** [comment]
However, I think it is a good candidate for feature request in any other general-purpose ML library like scikit-learn or special library for set of loses/metrics, not in LightGBM itself. I can't imagine one will install gradient boosting library with cpp code dependency only with the purpose of getting fast particular metric computation. Providing a separate dynamic library for only loses/metrics is not an option (at least not under this repository).

For the reference, source codes of loses/metrics:
- https://github.com/microsoft/LightGBM/tree/master/src/objective
- https://github.com/microsoft/LightGBM/tree/master/src/metric

**Comment by StrikerRUS** [comment]
I'd to hear any other opinion on whether this issue should go to feature requests.
@guolinke @jameslamb

**Comment by jameslamb** [comment]
> I can't imagine one will install gradient boosting library with cpp code dependency only with the purpose of getting fast particular metric computation

I agree, I don't think it's a good fit for LightGBM. It could be interesting to support a separate, standalone metrics library here but I don't think it's the most effective use of this small team of maintainers' limited time.

LightGBM uses a permissive license (MIT), so anyone contributing to say scikit-learn to speed up their metrics could take code from here and propose new code there based on it, as long as the LightGBM copyright is kept in a comment.

**Comment by wenmin-wu** [comment]
Thanks, @StrikerRUS and @jameslamb, let me forward this issue to scikit-learn.
